File: sdk/typescript/src/models-info.ts
/**
 * Model information database — pricing, context windows, capabilities.
 *
 * @module
 */

export interface ModelInfo {
  maxTokens: number;
  maxOutputTokens: number | null;
  inputCostPerToken: number;
  outputCostPerToken: number;
  supportsVision: boolean;
  supportsFunctionCalling: boolean;
  supportsJsonMode: boolean;
}

function m(
  maxTokens: number,
  opts: Partial<Omit<ModelInfo, "maxTokens">> = {},
): ModelInfo {
  return {
    maxTokens,
    maxOutputTokens: opts.maxOutputTokens ?? null,
    inputCostPerToken: opts.inputCostPerToken ?? 0,
    outputCostPerToken: opts.outputCostPerToken ?? 0,
    supportsVision: opts.supportsVision ?? false,
    supportsFunctionCalling: opts.supportsFunctionCalling ?? false,
    supportsJsonMode: opts.supportsJsonMode ?? false,
  };
}

// Pricing as of Feb 2026 — update periodically.
const MODEL_INFO: Record<string, ModelInfo> = {
  // OpenAI
  "gpt-4o": m(128000, {
    maxOutputTokens: 16384,
    inputCostPerToken: 2.5e-6,
    outputCostPerToken: 10e-6,
    supportsVision: true,
    supportsFunctionCalling: true,
    supportsJsonMode: true,
  }),
  "gpt-4o-mini": m(128000, {
    maxOutputTokens: 16384,
    inputCostPerToken: 0.15e-6,
    outputCostPerToken: 0.6e-6,
    supportsVision: true,
    supportsFunctionCalling: true,
    supportsJsonMode: true,
  }),
  "gpt-4-turbo": m(128000, {
    maxOutputTokens: 4096,
    inputCostPerToken: 10e-6,
    outputCostPerToken: 30e-6,
    supportsVision: true,
    supportsFunctionCalling: true,
    supportsJsonMode: true,
  }),
  "gpt-4": m(8192, {
    maxOutputTokens: 8192,
    inputCostPerToken: 30e-6,
    outputCostPerToken: 60e-6,
    supportsFunctionCalling: true,
  }),
  "gpt-3.5-turbo": m(16385, {
    maxOutputTokens: 4096,
    inputCostPerToken: 0.5e-6,
    outputCostPerToken: 1.5e-6,
    supportsFunctionCalling: true,
    supportsJsonMode: true,
  }),
  o1: m(200000, {
    maxOutputTokens: 100000,
    inputCostPerToken: 15e-6,
    outputCostPerToken: 60e-6,
    supportsVision: true,
    supportsFunctionCalling: true,
  }),
  "o1-mini": m(128000, {
    maxOutputTokens: 65536,
    inputCostPerToken: 3e-6,
    outputCostPerToken: 12e-6,
    supportsFunctionCalling: true,
  }),
  "o3-mini": m(200000, {
    maxOutputTokens: 100000,
    inputCostPerToken: 1.1e-6,
    outputCostPerToken: 4.4e-6,
    supportsFunctionCalling: true,
  }),

  // OpenAI Embeddings
  "text-embedding-3-small": m(8191, { inputCostPerToken: 0.02e-6 }),
  "text-embedding-3-large": m(8191, { inputCostPerToken: 0.13e-6 }),
  "text-embedding-ada-002": m(8191, { inputCostPerToken: 0.1e-6 }),

  // Anthropic
  "claude-sonnet-4-20250514": m(200000, {
    maxOutputTokens: 64000,
    inputCostPerToken: 3e-6,
    outputCostPerToken: 15e-6,
    supportsVision: true,
    supportsFunctionCalling: true,
    supportsJsonMode: true,
  }),
  "claude-3-5-sonnet-20241022": m(200000, {
    maxOutputTokens: 8192,
    inputCostPerToken: 3e-6,
    outputCostPerToken: 15e-6,
    supportsVision: true,
    supportsFunctionCalling: true,
    supportsJsonMode: true,
  }),
  "claude-3-5-haiku-20241022": m(200000, {
    maxOutputTokens: 8192,
    inputCostPerToken: 0.8e-6,
    outputCostPerToken: 4e-6,
    supportsFunctionCalling: true,
    supportsJsonMode: true,
  }),
  "claude-3-opus-20240229": m(200000, {
    maxOutputTokens: 4096,
    inputCostPerToken: 15e-6,
    outputCostPerToken: 75e-6,
    supportsVision: true,
    supportsFunctionCalling: true,
  }),
  "claude-3-haiku-20240307": m(200000, {
    maxOutputTokens: 4096,
    inputCostPerToken: 0.25e-6,
    outputCostPerToken: 1.25e-6,
    supportsVision: true,
    supportsFunctionCalling: true,
  }),

  // Google
  "gemini-2.0-flash": m(1048576, {
    maxOutputTokens: 8192,
    inputCostPerToken: 0.1e-6,
    outputCostPerToken: 0.4e-6,
    supportsVision: true,
    supportsFunctionCalling: true,
    supportsJsonMode: true,
  }),
  "gemini-1.5-pro": m(2097152, {
    maxOutputTokens: 8192,
    inputCostPerToken: 1.25e-6,
    outputCostPerToken: 5e-6,
    supportsVision: true,
    supportsFunctionCalling: true,
    supportsJsonMode: true,
  }),
  "gemini-1.5-flash": m(1048576, {
    maxOutputTokens: 8192,
    inputCostPerToken: 0.075e-6,
    outputCostPerToken: 0.3e-6,
    supportsVision: true,
    supportsFunctionCalling: true,
    supportsJsonMode: true,
  }),

  // Meta
  "llama-3.1-70b": m(128000, {
    maxOutputTokens: 4096,
    inputCostPerToken: 0.88e-6,
    outputCostPerToken: 0.88e-6,
    supportsFunctionCalling: true,
  }),
  "llama-3.1-8b": m(128000, {
    maxOutputTokens: 4096,
    inputCostPerToken: 0.18e-6,
    outputCostPerToken: 0.18e-6,
  }),

  // Mistral
  "mistral-large-latest": m(128000, {
    maxOutputTokens: 4096,
    inputCostPerToken: 2e-6,
    outputCostPerToken: 6e-6,
    supportsFunctionCalling: true,
    supportsJsonMode: true,
  }),
  "mistral-small-latest": m(128000, {
    maxOutputTokens: 4096,
    inputCostPerToken: 0.2e-6,
    outputCostPerToken: 0.6e-6,
    supportsFunctionCalling: true,
    supportsJsonMode: true,
  }),
};

/**
 * Map short names to full model names.
 * E.g. `modelAliasMap["gpt4"] = "gpt-4o"`.
 */
export const modelAliasMap: Record<string, string> = {};

// ---------------------------------------------------------------------------
// Lookups
// ---------------------------------------------------------------------------

/**
 * Look up model info by name.
 *
 * Resolution order:
 * 1. Exact match.
 * 2. Alias lookup via `modelAliasMap`.
 * 3. Prefix match (e.g. `"gpt-4o-2024-08-06"` → `"gpt-4o"`).
 */
export function getModelInfo(model: string): ModelInfo | null {
  // 1. Exact match
  if (model in MODEL_INFO) return MODEL_INFO[model];

  // 2. Alias resolution
  const resolved = modelAliasMap[model];
  if (resolved && resolved in MODEL_INFO) return MODEL_INFO[resolved];

  // 3. Prefix match (longest match first)
  const sortedKeys = Object.keys(MODEL_INFO).sort(
    (a, b) => b.length - a.length,
  );
  for (const known of sortedKeys) {
    if (model.startsWith(known)) return MODEL_INFO[known];
  }

  return null;
}

/** Return all known model names. */
export function getValidModels(): string[] {
  return Object.keys(MODEL_INFO);
}

/** Register or update a model in the info database. */
export function registerModel(modelName: string, info: ModelInfo): void {
  MODEL_INFO[modelName] = info;
}

// ---------------------------------------------------------------------------
// Capability checks
// ---------------------------------------------------------------------------

export function supportsVision(model: string): boolean {
  return getModelInfo(model)?.supportsVision ?? false;
}

export function supportsFunctionCalling(model: string): boolean {
  return getModelInfo(model)?.supportsFunctionCalling ?? false;
}

export function supportsJsonMode(model: string): boolean {
  return getModelInfo(model)?.supportsJsonMode ?? false;
}

export function supportsResponseSchema(model: string): boolean {
  return supportsJsonMode(model);
}

// ---------------------------------------------------------------------------
// Environment validation
// ---------------------------------------------------------------------------

const REQUIRED_ENV_VARS = ["AGENTCC_API_KEY", "AGENTCC_BASE_URL"];

export function validateEnvironment(): {
  keysSet: string[];
  keysMissing: string[];
  ready: boolean;
} {
  const keysSet: string[] = [];
  const keysMissing: string[] = [];

  for (const v of REQUIRED_ENV_VARS) {
    const val =
      typeof process !== "undefined" ? process.env?.[v] : undefined;
    if (val) {
      keysSet.push(v);
    } else {
      keysMissing.push(v);
    }
  }
  return { keysSet, keysMissing, ready: keysMissing.length === 0 };
}

File: sdk/python/src/agentcc/_models_info.py
"""Model information database -- pricing, context windows, capabilities."""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class ModelInfo:
    """Metadata about an LLM model."""

    max_tokens: int
    max_output_tokens: int | None = None
    input_cost_per_token: float = 0.0  # USD
    output_cost_per_token: float = 0.0  # USD
    supports_vision: bool = False
    supports_function_calling: bool = False
    supports_json_mode: bool = False


# Pricing as of Feb 2026 -- update periodically
MODEL_INFO: dict[str, ModelInfo] = {
    # OpenAI
    "gpt-4o": ModelInfo(
        max_tokens=128000,
        max_output_tokens=16384,
        input_cost_per_token=2.5e-6,
        output_cost_per_token=10e-6,
        supports_vision=True,
        supports_function_calling=True,
        supports_json_mode=True,
    ),
    "gpt-4o-mini": ModelInfo(
        max_tokens=128000,
        max_output_tokens=16384,
        input_cost_per_token=0.15e-6,
        output_cost_per_token=0.6e-6,
        supports_vision=True,
        supports_function_calling=True,
        supports_json_mode=True,
    ),
    "gpt-4-turbo": ModelInfo(
        max_tokens=128000,
        max_output_tokens=4096,
        input_cost_per_token=10e-6,
        output_cost_per_token=30e-6,
        supports_vision=True,
        supports_function_calling=True,
        supports_json_mode=True,
    ),
    "gpt-4": ModelInfo(
        max_tokens=8192,
        max_output_tokens=8192,
        input_cost_per_token=30e-6,
        output_cost_per_token=60e-6,
        supports_function_calling=True,
    ),
    "gpt-3.5-turbo": ModelInfo(
        max_tokens=16385,
        max_output_tokens=4096,
        input_cost_per_token=0.5e-6,
        output_cost_per_token=1.5e-6,
        supports_function_calling=True,
        supports_json_mode=True,
    ),
    "o1": ModelInfo(
        max_tokens=200000,
        max_output_tokens=100000,
        input_cost_per_token=15e-6,
        output_cost_per_token=60e-6,
        supports_vision=True,
        supports_function_calling=True,
    ),
    "o1-mini": ModelInfo(
        max_tokens=128000,
        max_output_tokens=65536,
        input_cost_per_token=3e-6,
        output_cost_per_token=12e-6,
        supports_function_calling=True,
    ),
    "o3-mini": ModelInfo(
        max_tokens=200000,
        max_output_tokens=100000,
        input_cost_per_token=1.1e-6,
        output_cost_per_token=4.4e-6,
        supports_function_calling=True,
    ),
    # OpenAI Embeddings
    "text-embedding-3-small": ModelInfo(
        max_tokens=8191,
        input_cost_per_token=0.02e-6,
    ),
    "text-embedding-3-large": ModelInfo(
        max_tokens=8191,
        input_cost_per_token=0.13e-6,
    ),
    "text-embedding-ada-002": ModelInfo(
        max_tokens=8191,
        input_cost_per_token=0.1e-6,
    ),
    # Anthropic
    "claude-sonnet-4-20250514": ModelInfo(
        max_tokens=200000,
        max_output_tokens=64000,
        input_cost_per_token=3e-6,
        output_cost_per_token=15e-6,
        supports_vision=True,
        supports_function_calling=True,
        supports_json_mode=True,
    ),
    "claude-3-5-sonnet-20241022": ModelInfo(
        max_tokens=200000,
        max_output_tokens=8192,
        input_cost_per_token=3e-6,
        output_cost_per_token=15e-6,
        supports_vision=True,
        supports_function_calling=True,
        supports_json_mode=True,
    ),
    "claude-3-5-haiku-20241022": ModelInfo(
        max_tokens=200000,
        max_output_tokens=8192,
        input_cost_per_token=0.8e-6,
        output_cost_per_token=4e-6,
        supports_function_calling=True,
        supports_json_mode=True,
    ),
    "claude-3-opus-20240229": ModelInfo(
        max_tokens=200000,
        max_output_tokens=4096,
        input_cost_per_token=15e-6,
        output_cost_per_token=75e-6,
        supports_vision=True,
        supports_function_calling=True,
    ),
    "claude-3-haiku-20240307": ModelInfo(
        max_tokens=200000,
        max_output_tokens=4096,
        input_cost_per_token=0.25e-6,
        output_cost_per_token=1.25e-6,
        supports_vision=True,
        supports_function_calling=True,
    ),
    # Google
    "gemini-2.0-flash": ModelInfo(
        max_tokens=1048576,
        max_output_tokens=8192,
        input_cost_per_token=0.1e-6,
        output_cost_per_token=0.4e-6,
        supports_vision=True,
        supports_function_calling=True,
        supports_json_mode=True,
    ),
    "gemini-1.5-pro": ModelInfo(
        max_tokens=2097152,
        max_output_tokens=8192,
        input_cost_per_token=1.25e-6,
        output_cost_per_token=5e-6,
        supports_vision=True,
        supports_function_calling=True,
        supports_json_mode=True,
    ),
    "gemini-1.5-flash": ModelInfo(
        max_tokens=1048576,
        max_output_tokens=8192,
        input_cost_per_token=0.075e-6,
        output_cost_per_token=0.3e-6,
        supports_vision=True,
        supports_function_calling=True,
        supports_json_mode=True,
    ),
    # Meta
    "llama-3.1-70b": ModelInfo(
        max_tokens=128000,
        max_output_tokens=4096,
        input_cost_per_token=0.88e-6,
        output_cost_per_token=0.88e-6,
        supports_function_calling=True,
    ),
    "llama-3.1-8b": ModelInfo(
        max_tokens=128000,
        max_output_tokens=4096,
        input_cost_per_token=0.18e-6,
        output_cost_per_token=0.18e-6,
    ),
    # Mistral
    "mistral-large-latest": ModelInfo(
        max_tokens=128000,
        max_output_tokens=4096,
        input_cost_per_token=2e-6,
        output_cost_per_token=6e-6,
        supports_function_calling=True,
        supports_json_mode=True,
    ),
    "mistral-small-latest": ModelInfo(
        max_tokens=128000,
        max_output_tokens=4096,
        input_cost_per_token=0.2e-6,
        output_cost_per_token=0.6e-6,
        supports_function_calling=True,
        supports_json_mode=True,
    ),
}

model_alias_map: dict[str, str] = {}
"""Map short names to full model names. E.g. ``{"gpt4": "gpt-4o"}``"""

# --- Required environment variables ---

_REQUIRED_ENV_VARS: list[str] = ["AGENTCC_API_KEY", "AGENTCC_BASE_URL"]


def get_model_info(model: str) -> ModelInfo | None:
    """Look up model info by name.

    Resolution order:

    1. Exact match in ``MODEL_INFO``.
    2. Alias lookup via ``model_alias_map``, then exact match.
    3. Prefix match (e.g., ``"gpt-4o-2024-08-06"`` -> ``"gpt-4o"``).
    """
    # 1. Exact match
    if model in MODEL_INFO:
        return MODEL_INFO[model]

    # 2. Alias resolution
    resolved = model_alias_map.get(model)
    if resolved is not None and resolved in MODEL_INFO:
        return MODEL_INFO[resolved]

    # 3. Prefix match (longest match first)
    for known_model in sorted(MODEL_INFO.keys(), key=len, reverse=True):
        if model.startswith(known_model):
            return MODEL_INFO[known_model]
    return None


def get_valid_models() -> list[str]:
    """Return all known model names from the model info database."""
    return list(MODEL_INFO.keys())


def register_model(model_name: str, model_info: ModelInfo) -> None:
    """Register or update a model in the info database.

    Args:
        model_name: The model identifier (e.g. ``"my-custom-model"``).
        model_info: A :class:`ModelInfo` instance with the model's metadata.
    """
    MODEL_INFO[model_name] = model_info


def supports_vision(model: str) -> bool:
    """Check if *model* supports vision/image inputs."""
    info = get_model_info(model)
    return info.supports_vision if info else False


def supports_function_calling(model: str) -> bool:
    """Check if *model* supports function/tool calling."""
    info = get_model_info(model)
    return info.supports_function_calling if info else False


def supports_json_mode(model: str) -> bool:
    """Check if *model* supports JSON mode / structured output."""
    info = get_model_info(model)
    return info.supports_json_mode if info else False


def supports_response_schema(model: str) -> bool:
    """Alias for :func:`supports_json_mode`."""
    return supports_json_mode(model)


def validate_environment() -> dict[str, Any]:
    """Check if required environment variables are set.

    Returns:
        A dict with:
            - ``keys_set``: list of env var names that are set.
            - ``keys_missing``: list of env var names that are missing.
            - ``ready``: ``True`` if all required vars are set.
    """
    keys_set: list[str] = []
    keys_missing: list[str] = []

    for var in _REQUIRED_ENV_VARS:
        if os.environ.get(var):
            keys_set.append(var)
        else:
            keys_missing.append(var)

    return {
        "keys_set": keys_set,
        "keys_missing": keys_missing,
        "ready": len(keys_missing) == 0,
    }
